pmlc/dialect/tile/transforms/pad_constraints.cc

```diff
@@ -0,0 +1,266 @@
+        continue;
+      // Get the tensor and the map associated with the tensor
+      auto ttype = in.getType().cast<mlir::TensorType>();
+      auto map = op.getSourceMap(i);
+      // The should have the same rank.
+      assert(ttype.getRank() == map.getNumResults());
+      // Add all of the constraints that define the valid region of a tensor
+      inputs.emplace_back();
+      auto &cons = inputs.back();
+      for (int64_t dim = 0; dim < ttype.getRank(); dim++) {
+        // Get dimension specific expression
+        auto expr = map.getResult(dim);
+        // Lower bound (expr >= 0)
+        cons.emplace_back(expr);
+        // Upper bound (expr <= (size - 1) or in canonical form:
+        // (size - 1) - expr >= 0
+        cons.emplace_back(-expr + (ttype.getDimSize(dim) - 1));
+      }
+    }
+    // If no padded inputs, early exit
+    if (inputs.empty())
+      return;
+    // Check each constraint by adding its inverse to each input polyhedra.
+    // If the resulting integer set is empty, then all of the false cases are
+    // outside of the valid region (i.e. in the padded region).  Thus any loads
+    // when the constraint is false will load the padding value, and thus have
+    // no effect on the output.  This means that removing the constraint is
+    // safe, since all places it would have previously prevented will turn into
+    // no-ops post padding.
+    for (auto expr : op.cons()->getConstraints()) {
+      // Compute the 'inverse' of the constraint.
+      // The opposite of expr >= 0 is expr < 0
+      // Given expr is an integer this is the same as expr + 1 <= 0 or
+      // -expr - 1 >= 0
+      auto inv_expr = -expr - 1;
+      // Initially presume we will not keep this constraint
+      bool keep = true;
+      // For each input
+      for (auto &cons : inputs) {
+        // Temporarily add the inv_expr to the constraints
+        cons.push_back(inv_expr);
+        // Make into an IntegerSet, and then into FlatAffineConstraints.
+        // It seems wasteful to intern a temporary integer set, but any other
+        // way of doing this is also annoying given the current class structures
+        auto set = makeConstraintSet(numDims, cons);
+        mlir::FlatAffineConstraints fac(set);
+        // Can't prove it's empty, keep constraint
+        if (fac.isEmpty())
+          keep = false;
+        // Remove
+        cons.pop_back();
+        // If we've decided not to keep the constraint, don't bother considering
+        // more inputs.
+        if (!keep) {
+          break;
+        }
+      }
+      if (keep)
+        savedConstraints.push_back(expr);
+    }
+    if (savedConstraints.empty()) {
+      op.removeAttr(ContractionOp::getConstraintsAttrName());
+    } else {
+      op.setConstraints(makeConstraintSet(numDims, savedConstraints));
+    }
+  });
+}
+
+} // namespace
+
+llvm::Optional<PaddingInfo> getPaddingInfo(Operation *op) {
+  if (!op)
+    return llvm::None;
```

Comment:
Would it be useful to have error messages with these failure cases? Or is it OK/expected to fail in some cases?

Reply:
Yes, in general the goal of this code is to extract the padding attributes back into a c++ structure if they exist, but in most cases (non-padded operations) they will not.  To be honest, we probably should use a custom attribute here, but given that this is really just a moved copy of the existing padding pass, I don't think it's worth mucking with right now.